runners/google-cloud-dataflow-java/worker/src/test/java/org/apache/beam/runners/dataflow/worker/windmill/client/grpc/GrpcDispatcherClientTest.java

```diff
@@ -0,0 +1,159 @@
+import org.apache.beam.sdk.options.PipelineOptionsFactory;
+import org.apache.beam.vendor.grpc.v1p60p1.io.grpc.testing.GrpcCleanupRule;
+import org.apache.beam.vendor.guava.v32_1_2_jre.com.google.common.collect.ImmutableSet;
+import org.apache.beam.vendor.guava.v32_1_2_jre.com.google.common.collect.Lists;
+import org.apache.beam.vendor.guava.v32_1_2_jre.com.google.common.net.HostAndPort;
+import org.junit.Rule;
+import org.junit.Test;
+import org.junit.experimental.runners.Enclosed;
+import org.junit.runner.RunWith;
+import org.junit.runners.JUnit4;
+import org.junit.runners.Parameterized;
+import org.junit.runners.Parameterized.Parameter;
+import org.junit.runners.Parameterized.Parameters;
+
+@RunWith(Enclosed.class)
+public class GrpcDispatcherClientTest {
+
+  static class GrpcDispatcherClientTestBase {
+
+    @Rule public final GrpcCleanupRule grpcCleanup = new GrpcCleanupRule();
+
+    final ChannelCachingStubFactory stubFactory =
+        new FakeWindmillStubFactory(
+            () ->
+                grpcCleanup.register(
+                    WindmillChannelFactory.inProcessChannel("GrpcDispatcherClientTestChannel")));
+
+    static StreamingGlobalConfig getGlobalConfig(boolean useWindmillIsolatedChannels) {
+      return StreamingGlobalConfig.builder()
+          .setWindmillServiceEndpoints(ImmutableSet.of(HostAndPort.fromString("windmill:1234")))
+          .setUserWorkerJobSettings(
+              UserWorkerRunnerV1Settings.newBuilder()
+                  .setUseWindmillIsolatedChannels(useWindmillIsolatedChannels)
+                  .build())
+          .build();
+    }
+  }
+
+  @RunWith(JUnit4.class)
+  public static class RespectsJobSettingTest extends GrpcDispatcherClientTestBase {
+
+    @Test
+    public void createsNewStubWhenIsolatedChannelsConfigIsChanged() {
+      DataflowWorkerHarnessOptions options =
+          PipelineOptionsFactory.as(DataflowWorkerHarnessOptions.class);
+      options.setExperiments(
+          Lists.newArrayList(
+              GrpcDispatcherClient.STREAMING_ENGINE_USE_JOB_SETTINGS_FOR_ISOLATED_CHANNELS));
+      GrpcDispatcherClient dispatcherClient =
+          GrpcDispatcherClient.create(options, new FakeWindmillStubFactoryFactory(stubFactory));
+      // Create first time with Isolated channels disabled
+      dispatcherClient.onJobConfig(getGlobalConfig(/*useWindmillIsolatedChannels=*/ false));
+      CloudWindmillServiceV1Alpha1Stub stub1 = dispatcherClient.getWindmillServiceStub();
+      CloudWindmillServiceV1Alpha1Stub stub2 = dispatcherClient.getWindmillServiceStub();
+      assertSame(stub2, stub1);
+
+      // Enable Isolated channels
+      dispatcherClient.onJobConfig(getGlobalConfig(/*useWindmillIsolatedChannels=*/ true));
+      CloudWindmillServiceV1Alpha1Stub stub3 = dispatcherClient.getWindmillServiceStub();
+      assertNotSame(stub3, stub1);
+
+      CloudWindmillServiceV1Alpha1Stub stub4 = dispatcherClient.getWindmillServiceStub();
+      assertSame(stub3, stub4);
```

Comment:
this doesn't seem to check the isolation is enabled or not, just that the stubs were recreated

Reply:
Fixed.